src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -520,25 +520,41 @@ protected ComputedInstanceFieldLayout ComputeAutoFieldLayout(MetadataType type,
                 if (!fieldLayoutAbiStable)
                     layoutAbiStable = false;
 
-                largestAlignmentRequired = LayoutInt.Max(fieldSizeAndAlignment.Alignment, largestAlignmentRequired);
-
                 if (IsByValueClass(fieldType))
                 {
+                    largestAlignmentRequired = LayoutInt.Max(fieldSizeAndAlignment.Alignment, largestAlignmentRequired);
                     instanceValueClassFieldsArr[instanceValueClassFieldCount++] = field;
                 }
-                else if (fieldType.IsGCPointer)
-                {
-                    instanceGCPointerFieldsArr[instanceGCPointerFieldsCount++] = field;
-                }
                 else
                 {
-                    int log2size = CalculateLog2(fieldSizeAndAlignment.Size.AsInt);
-                    instanceNonGCPointerFieldsArr[log2size][instanceNonGCPointerFieldsCount[log2size]++] = field;
+                    // non-value-type fields always require pointer alignment
+                    // This does not account for types that are marked IsAlign8Candidate due to 8-byte fields
+                    // but that is explicitly handled when we calculate the final alignment for the type.
+
+                    // This behavior is extremely strange for primitive types, as it makes a struct with a single byte in it
```

Comment:
Can you elaborate on this? Wouldn't `struct S { byte x; }` hit `IsByValueClass(fieldType)` above and so it would be `1-byte alignment`?

Should this instead say `class` since it looks like `class C { byte x; }` would hit this case and it would be `align: sizeof(void*)`?

Reply:
Hmm, these questions need David's response but he's on vacation next week. Should we hold off merging this change until he's back so that he can answer them?

Reply:
Was this targeting .NET 7?

If so then for this one, its just a question on comment wording, I don't think it's worth blocking over.

For the one below, I think it's worth confirming that CG2 is at least matching what the RyuJIT VM does. If it is, then I think its fine to merge.

Reply:
The comment needs clarity, and possibly a test about this particularly odd behavior. From my reading, the rules are that:
```
struct S1
{
    byte b;
}
[StructLayout(LayoutKind.Auto)]
struct S1_Auto
{
    byte b;
}
struct S2
{
    S1 fld1;
    S1_Auto fld2;
}
```

Assuming 64bit...
This will result in `S1` having size of 1, and alignment of 1, `S1_Auto` having size of 8 and alignment of 8, and `S2` having size of 16, and alignment of 8. Note that this wacky behavior is exclusive to auto layout structures.

Largest alignment rules for classes are less interesting in general, and pretty much just impact ARM32, where a class alignment of 8 is significant, and triggers the 8 byte alignment behavior. However, we still calculate them, in case we add support for 16 byte or higher object alignment rules in the future.

Reply:
Also `IsByValueClass(t)` is false for primitive and enum types, and so doesn't go into the structure alignment path.

Reply:
> Note that this wacky behavior is exclusive to auto layout structures.

I thought we fixed some of this for `generics` so that `Nullable<T>` wasn't "too large" anymore? Or was that a different fix since its generic but sequential?

This, more generally, seems like something which may have undesirable consequences and which we may want to look at more in depth for .NET 8. Consider types like `ValueTuple` which are frequently used for tuples of primitives and where its marked `Auto` with the intent that the BCL optimizes the layout to be more efficient. If the alignment of fields is treated as `8`, that is not going to be the case and many usages will end up more expensive instead.

Reply:
Mmm... maybe? I've been writing tests, and to get the wacky layout behavior, I need to make the auto layout structure be larger than a pointer, so the tests that cover these boundary conditions are even dumber than I thought.